libs/utils/src/vec_map.rs

```diff
@@ -1,35 +1,64 @@
 use std::{alloc::Layout, cmp::Ordering, ops::RangeBounds};
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum VecMapOrdering {
+    Greater,
+    GreaterOrEqual,
+}
+
 /// Ordered map datastructure implemented in a Vec.
 /// Append only - can only add keys that are larger than the
 /// current max key.
 #[derive(Clone, Debug)]
-pub struct VecMap<K, V>(Vec<(K, V)>);
+pub struct VecMap<K, V> {
+    data: Vec<(K, V)>,
+    ordering: VecMapOrdering,
+}
 
 impl<K, V> Default for VecMap<K, V> {
     fn default() -> Self {
-        VecMap(Default::default())
+        VecMap {
+            data: Default::default(),
+            ordering: VecMapOrdering::Greater,
+        }
     }
 }
 
 #[derive(Debug)]
-pub struct InvalidKey;
+pub enum VecMapError {
+    InvalidKey,
+    ExtendOrderingError,
+}
```

Comment:
Let's use `thiserror` since we are turning this into a legit error type:
```
#[derive(thiserror::Error, Debug)]
pub enum VecMapError {
  #[error("Key violates ordering constraint")]
  InvalidKey,
  #[error("Mismatched ordering constraints")]
  ExtendOrderingError
};
```